pkg/retriever/retrieve.go

```diff
@@ -216,36 +223,106 @@ func collectResults(ctx context.Context, retrieval *retrieval, candidates []Retr
 // runRetrievalCandidate is a singular CID:SP retrieval, expected to be run in a goroutine
 // and coordinate with other candidate retrievals to block after query phase and
 // only attempt one retrieval-proper at a time.
-func runRetrievalCandidate(ctx context.Context, cfg *RetrievalConfig, client RetrievalClient, retrieval *retrieval, candidate RetrievalCandidate) {
+func runRetrievalCandidate(ctx context.Context, cfg *RetrievalConfig, client RetrievalClient, retrieval *retrieval, queryStartTime time.Time, candidate types.RetrievalCandidate) {
 	var timeout time.Duration
 	if cfg.GetStorageProviderTimeout != nil {
 		timeout = cfg.GetStorageProviderTimeout(candidate.MinerPeer.ID)
 	}
 
-	var stats *RetrievalStats
+	var stats *types.RetrievalStats
+	var retrievalErr error
 	var done func()
-	queryStartTime := time.Now()
+
+	retrieval.sendEvent(eventpublisher.Started(queryStartTime, eventpublisher.QueryPhase, candidate))
 
 	// run the query phase
-	queryResponse, err := queryPhase(ctx, cfg, client, timeout, candidate)
+	onConnected := func() {
+		retrieval.sendEvent(eventpublisher.Connect(queryStartTime, eventpublisher.QueryPhase, candidate))
+	}
+	queryResponse, queryErr := queryPhase(ctx, cfg, client, timeout, candidate, onConnected)
+	if queryErr != nil {
+		retrieval.sendEvent(eventpublisher.Failure(queryStartTime, eventpublisher.QueryPhase, candidate, queryErr.Error()))
+	}
 
 	if queryResponse != nil {
+		retrieval.sendEvent(eventpublisher.QueryAsk(queryStartTime, candidate, *queryResponse))
+		if queryResponse.Status != retrievalmarket.QueryResponseAvailable ||
+			(cfg.IsAcceptableQueryResponse != nil && !cfg.IsAcceptableQueryResponse(queryResponse)) {
+			queryResponse = nil
+		}
+	}
+
+	if queryResponse != nil {
+		retrieval.sendEvent(eventpublisher.QueryAskFiltered(queryStartTime, candidate, *queryResponse))
+
 		// if query is successful, then wait for priority and execute retrieval
 		done = retrieval.waitQueue.Wait(&queryCandidate{queryResponse, time.Since(queryStartTime)})
+		// only one goroutine is allowed to execute past here at a time, so retrieval.retrievalStartTime
+		// doesn't need to be protected by a mutex, but it should mark the time the first retrieval
+		// starts
+		if retrieval.retrievalStartTime.IsZero() {
+			retrieval.retrievalStartTime = time.Now()
+		}
```

Comment:
DON'T MERGE this branch yet

I have a race here that I need to work out - I need to figure out why this is a problem, it _shouldn't_ be (see comment above) so I'm missing something important here